### ISSUE TITLE
open spreadsheet by key/url THEN access worksheet by ID/GID

### ISSUE THREAD

**Issue by joaocasarin** [burnash/gspread#830]
Hello, I work with google spreadsheets, and inside them there are A LOT of worksheets (over 500), and sometimes it happens that some of them get the same title name, so the only mean to differ both, is by the GID, provided at the END of the URL like this:

https://docs.google.com/spreadsheets/d/1231SDFS3afer3SDfsdf-34SDFSDgfdsd42/edit#gid=5439787534

where:
1231SDFS3afer3SDfsdf-34SDFSDgfdsd42 ------ SPREADSHEET key/id
gid=5439787534 --------------------------------- WORKSHEET(the real place where cells are) ID

This is really urgent for me, as I need to fasten the time consumed by our job, so being able to access it directly by the ID would be by far one of the best things I could find on internet...

Thanks

**Comment by joaocasarin** [comment]
@burnash I'm sorry for mentioning you, but any help is appreciated, and this seems urgent for me :(

**Comment by a-kochkarev** [comment]
Hello! I think it will be nice if _gspread.models.Spreadsheet_ will have method like _worksheet_by_id_

**Comment by a-kochkarev** [comment]
> Hello, I work with google spreadsheets, and inside them there are A LOT of worksheets (over 500), and sometimes it happens that some of them get the same title name, so the only mean to differ both, is by the GID, provided at the END of the URL like this:
> 
> https://docs.google.com/spreadsheets/d/1231SDFS3afer3SDfsdf-34SDFSDgfdsd42/edit#gid=5439787534
> 
> where:
> 1231SDFS3afer3SDfsdf-34SDFSDgfdsd42 ------ SPREADSHEET key/id
> gid=5439787534 --------------------------------- WORKSHEET(the real place where cells are) ID
> 
> This is really urgent for me, as I need to fasten the time consumed by our job, so being able to access it directly by the ID would be by far one of the best things I could find on internet...
> 
> Thanks

Hello, @joaocasarin!

Until gspread not provide something like  _worksheet_by_id_ you can get worksheet by id with this sample of code:

```
import re

import gspread
from gspread.exceptions import WorksheetNotFound
from gspread.utils import finditem
from gspread.models import Worksheet


URL_SHEET_ID_RE = re.compile(r'#gid=([0-9]+)')

def extract_sheet_id_from_url(url):
    res = URL_SHEET_ID_RE.search(url)
    if res:
        return res.group(1)

    raise Exception('No valid sheet id found in url')


def open_worksheet_by_id(spreadsheet, sheetId):
    sheet_data = spreadsheet.fetch_sheet_metadata()
    try:
        item = finditem(
            lambda x: x['properties']['sheetId'] == int(sheetId),
            sheet_data['sheets'],
        )
        return Worksheet(spreadsheet, item['properties'])
    except (StopIteration, KeyError):
        raise WorksheetNotFound(sheetId)


SAMPLE_SPREADSHEET_URL = 'https://docs.google.com/spreadsheets/d/19d96IpQEvIj15Me0S9SApPgdHrWaZUKmtZE1kxFbdbq/edit#gid=1234123123'

def main():
    client = gspread.service_account('secrets\\service.json')
    sheet = client.open_by_url(SAMPLE_SPREADSHEET_URL)
    worksheet = open_worksheet_by_id(sheet, extract_sheet_id_from_url(SAMPLE_SPREADSHEET_URL))
    
    for row in worksheet.get_all_values():
         print(row)


if __name__ == '__main__':
    main()


```

**Comment by joaocasarin** [comment]
hello @a-kochkarev thanks for the help! Until the gspread devs don't provide such function, yours will fit perfectly! Thanks buddy!

**Comment by lavigne958** [comment]
This is a duplicate issue (see #271 ) and will be closed soon by #857 .

Sorry for the delay, thank you very much @akochkarev for the quick workaround.

As a duplicate, closing this issue.
